server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,178 @@
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(new Long(mail.getMessageSize()));
+        } catch (MessagingException e) {
+            throw new MissingRequestedField("messageSize");
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.BODY)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.ofNullable(mail.getMessage().getContent().toString());
+        } catch (IOException | MessagingException | NullPointerException e) {
+            throw new MissingRequestedField("body");
+        }
+    }
+
+    private static Optional<Map<String, List<String>>> fetchHeaders(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
+        if (!additionalFields.contains(AdditionalFields.HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, List<String>> headers = new HashMap<>();
+
+        Enumeration<Header> rawHeaders;
+        try {
+            rawHeaders = mail.getMessage().getAllHeaders();
+            while (rawHeaders.hasMoreElements()) {
+                Header header = rawHeaders.nextElement();
+                String name = header.getName();
+                if (!headers.containsKey(name)) {
+                    headers.put(name, new ArrayList<>());
+                }
+
+                headers.get(name).add(header.getValue());
+            }
+
+            return Optional.of(headers);
+        } catch (NullPointerException | MessagingException e) {
+            throw new MissingRequestedField("headers");
+        }
+    }
+
+    private static Optional<Map<String, Map<String, List<String>>>> fetchPerRecipientsHeaders(List<AdditionalFields> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalFields.PER_RECIPIENTS_HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, Map<String, List<String>>> headers = new HashMap<>();
+        PerRecipientHeaders specificHeaders = mail.getPerRecipientSpecificHeaders();
+        Multimap<MailAddress, org.apache.mailet.PerRecipientHeaders.Header> headersByRecipient = specificHeaders.getHeadersByRecipient();
+
+        for (MailAddress address : headersByRecipient.keySet()) {
+            headers.put(address.asString(), fetchPerRecipientHeader(headersByRecipient, address));
+        }
+
+        return Optional.of(headers);
+    }
+
+    private static Map<String, List<String>> fetchPerRecipientHeader(
+            Multimap<MailAddress, org.apache.mailet.PerRecipientHeaders.Header> headersByRecipient,
```

Comment:
you probably can make things simpler by calling ```headersByRecipient.get(address)``` outside of this method

Reply:
I did not receive explanation on why this FQDN is needed. Could you?

Reply:
There's not real reason, I'll remove it